### DISCUSSION TITLE
i need tap in a coordinate of my screen

### DISCUSSION THREAD

**Discussion by nimairodrigues** [appium/appium#18931] [Q&A]
i was trying to use the following method to tap in a coordinate, but its not working, idk why
```
public void tap(int x, int y) {
  new TouchAction<>(driver())
  .tap(PointOption.point(x, y))
  .perform();
}
```
Appium's log says:
```
[debug] [W3C] Matched W3C error code 'invalid element state' to InvalidElementStateError
[debug] [AndroidUiautomator2Driver@64d6 (7af3d052)] Encountered internal error running command: io.appium.uiautomator2.common.exceptions.InvalidElementStateException: Tap at [x=150.0, y=200.0] has failed
```

**Comment by jlipps** [comment]
Don't use `TouchAction`. instead use the w3c compatible actions class.

**Reply by nimairodrigues** [reply]
whats the problem? i use some codes like "scroll" and "swipe" using TouchAction, and works well.
i found this code below but it looks like a "click" and idk if has difference between click and tap, do you know?
```
public void tap(int x, int y) {
		PointerInput finger = new PointerInput(PointerInput.Kind.TOUCH, "finger");
		Sequence clickSequence = new Sequence(finger, 1);
		clickSequence.addAction(finger.createPointerMove(Duration.ofMillis(0), PointerInput.Origin.viewport(), x, y));
		clickSequence.addAction(finger.createPointerDown(PointerInput.MouseButton.LEFT.asArg()));
		getDriver().perform(Arrays.asList(clickSequence));
	}
```

**Reply by jlipps** [reply]
The problem is that `TouchAction` uses old and deprecated APIs. the new code snippet you pasted looks correct. There is no difference between click/tap. Change the variable names to `tapSequence` if you want.
